File: clients/typescript/src/ratelimiting/RateLimitHeaders.ts
import type { RateLimitPolicy } from "./RateLimitPolicy.js";
import type { RateLimitState } from "./RateLimitState.js";

/**
 * Parses the `RateLimit-Policy`, `RateLimit`, and `RateLimit-Cost` response headers.
 *
 * Header format:
 * ```
 * RateLimit-Policy: "burst";q=10000;w=300, "sustained";q=100000;w=2592000
 * RateLimit:        "burst";r=9990;t=300,  "sustained";r=99641;t=2589945
 * RateLimit-Cost:   10
 * ```
 */
export function parsePolicyHeader(header: string | null | undefined): RateLimitPolicy[] {
  return splitEntries(header)
    .map((entry) => {
      const tokens = tokensOf(entry);
      const name = tokens[0]?.replace(/^"|"$/g, "");
      const q = intParam(tokens, "q");
      const w = intParam(tokens, "w");
      if (!name || q === undefined || w === undefined) return null;
      return { name, quota: q, windowSeconds: w } satisfies RateLimitPolicy;
    })
    .filter((p): p is RateLimitPolicy => p !== null);
}

export function parseStateHeader(header: string | null | undefined): RateLimitState[] {
  return splitEntries(header)
    .map((entry) => {
      const tokens = tokensOf(entry);
      const name = tokens[0]?.replace(/^"|"$/g, "");
      const r = intParam(tokens, "r");
      const t = intParam(tokens, "t");
      if (!name || r === undefined || t === undefined) return null;
      return { name, remaining: r, resetSeconds: t } satisfies RateLimitState;
    })
    .filter((s): s is RateLimitState => s !== null);
}

export function parseCostHeader(header: string | null | undefined): number {
  if (!header) return 0;
  const n = parseInt(header.trim(), 10);
  return Number.isFinite(n) ? n : 0;
}

function splitEntries(header: string | null | undefined): string[] {
  if (!header || !header.trim()) return [];
  return header.split(",").map((s) => s.trim()).filter(Boolean);
}

function tokensOf(entry: string): string[] {
  return entry.split(";").map((s) => s.trim());
}

function intParam(tokens: string[], key: string): number | undefined {
  const prefix = `${key}=`;
  const token = tokens.slice(1).find((t) => t.startsWith(prefix));
  if (!token) return undefined;
  const n = parseInt(token.slice(prefix.length), 10);
  return Number.isFinite(n) ? n : undefined;
}

File: clients/swift/src/RateLimiting/RateLimitHeaders.swift
import Foundation

/// Parses the `RateLimit-Policy`, `RateLimit`, and `RateLimit-Cost` response headers.
///
/// Header format:
/// ```
/// RateLimit-Policy: "burst";q=10000;w=300, "sustained";q=100000;w=2592000
/// RateLimit:        "burst";r=9990;t=300,  "sustained";r=99641;t=2589945
/// RateLimit-Cost:   10
/// ```
public enum RateLimitHeaders {
    public static func parsePolicy(_ header: String?) -> [RateLimitPolicy] {
        splitEntries(header).compactMap { entry in
            let tokens = tokens(of: entry)
            guard let name = tokens.first.map({ $0.trimmingCharacters(in: .init(charactersIn: "\"")) }),
                  let q = intParam(tokens, key: "q"),
                  let w = intParam(tokens, key: "w")
            else { return nil }
            return RateLimitPolicy(name: name, quota: q, windowSeconds: w)
        }
    }

    public static func parseState(_ header: String?) -> [RateLimitState] {
        splitEntries(header).compactMap { entry in
            let tokens = tokens(of: entry)
            guard let name = tokens.first.map({ $0.trimmingCharacters(in: .init(charactersIn: "\"")) }),
                  let r = intParam(tokens, key: "r"),
                  let t = intParam(tokens, key: "t")
            else { return nil }
            return RateLimitState(name: name, remaining: r, resetSeconds: t)
        }
    }

    public static func parseCost(_ header: String?) -> Int {
        Int(header?.trimmingCharacters(in: .whitespaces) ?? "") ?? 0
    }

    // MARK: - Internals

    private static func splitEntries(_ header: String?) -> [String] {
        guard let header, !header.trimmingCharacters(in: .whitespaces).isEmpty else { return [] }
        return header.split(separator: ",").map { $0.trimmingCharacters(in: .whitespaces) }
    }

    private static func tokens(of entry: String) -> [String] {
        entry.split(separator: ";").map { $0.trimmingCharacters(in: .whitespaces) }
    }

    private static func intParam(_ tokens: [String], key: String) -> Int? {
        let prefix = "\(key)="
        guard let token = tokens.dropFirst().first(where: { $0.hasPrefix(prefix) }) else { return nil }
        return Int(token.dropFirst(prefix.count))
    }
}
